jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,149 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+
+    // Get the relevant events (events with at least one required attendee).
+    Collection<Event> relevantEvents = getRelevantEvents(events, requiredAttendees);
+
+    // Get the TimeRanges of the relevant events in a List, ordered by start time.
+    List<TimeRange> timeRanges = getEventTimeRanges(relevantEvents); 
+
+    // Get a condensed TimeRanges list for the relevent events (condensed means overlapping events are turned into a single event).
+    List<TimeRange> condensedTimeRanges = getCondensedTimeRanges(timeRanges);
+
+    // Get the inverse time ranges in a day, given a collection of non-overlapping time ranges
+    // sorted by their start time. Inverse means all the time ranges not covered by any time range
+    // in the collection.
+    List<TimeRange> openTimeRanges = getInverseTimeRanges(condensedTimeRanges);
+
+    // Filter the open time ranges such that only open time ranges with a duration greater than the duration of the
+    // meeting remain
+    List<TimeRange> validTimeRanges = getValidTimeRanges(openTimeRanges, (int) request.getDuration());
+
+    return validTimeRanges;
   }
+
+  // Adds all relevant events (events with at least one required attendee) to a new collection.
+  private Collection<Event> getRelevantEvents(Collection<Event> events, Collection<String> requiredAttendees) {
+    Collection<Event> relevantEvents = new HashSet<>();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+
+      if (!Collections.disjoint(requiredAttendees, eventAttendees)) {
+        relevantEvents.add(event);
+      }
+    }
+
+    return relevantEvents;
+  }
+
+  // Adds the timeranges of all events in the collection to a list, sorted by start time.
+  private List<TimeRange> getEventTimeRanges(Collection<Event> events) {
+    List<TimeRange> timeRanges = new ArrayList<>();
+
+    for (Event event : events) {
+      timeRanges.add(event.getWhen());
+    }
+
+    // Sort the timeRanges by start time.
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    return timeRanges;
+  }
+
+  /**
+   * Condenses an ordered list of time ranges (ordered by start time).
+   * @param timeRanges list of timeRanges sorted by start time
+   * @return list of condensed timeRanges (condensed means overlapping events are turned into a single event).
```

Comment:
Are adjacent events considered overlapping? e.g. would an 8:00->8:30 event be condensed with an 8:30->9:00 event? May be worth pointing this out in the docs.

Reply:
No - adjacent events are not condensed. This follows the definition of "overlapping timeRanges" from the timeRange class. I updated the documentation to include this information.